src/BridgeFactory.sol

```diff
@@ -36,70 +48,66 @@ contract BridgeFactory is IERC1271, MultiCallContext, TwoStepOwnable {
+    function _verifyRoot(bytes32 root, address pendingOwner_) internal view {
+        bytes32 initHash = _proxyInitHash;
+        address factory = _cachedThis;
+        assembly ("memory-safe") {
+            let ptr := mload(0x40)
+
+            // derive creation salt
+            mstore(0x00, root)
+            mstore(0x14, pendingOwner_)
+            mstore(0x34, chainid())
+            let salt := keccak256(0x00, 0x54)
+
+            // 0xff + factory + salt + hash(initCode)
+            mstore(0x0d, factory)
+            mstore(0x00, 0xff00000000000000)
+            mstore(0x2d, salt)
+            mstore(0x4d, initHash)
+            let computedAddress := keccak256(0x18, 0x55)
+
+            // verify that `salt` was used to deploy `address(this)`
+            if shl(0x60, xor(address(), computedAddress)) {
+                mstore(0x00, 0x1e092104) // selector for `PermissionDenied()`
+                revert(0x1c, 0x04)
+            }
+
+            // restore clobbered memory
+            mstore(0x40, ptr)
+        }
+    }
+
+    function isValidSignature(bytes32 hash, bytes calldata signature)
         external
         view
         override
         onlyProxy
         returns (bytes4)
     {
-        return _isValidSignature(_hash, _signature);
-    }
-
-    function _isValidSignature(bytes32 _leaf, bytes calldata _proof) private view returns (bytes4) {
         bytes32[] calldata proof;
         assembly ("memory-safe") {
-            // _signature is just going to be the proof, then we can read it as so
-            proof.offset := add(0x20, _proof.offset)
+            // signature is just the proof, then we can read it as so
+            proof.offset := add(signature.offset, calldataload(signature.offset))
             proof.length := calldataload(proof.offset)
             proof.offset := add(0x20, proof.offset)
         }
-        bytes32 salt = keccak256(abi.encodePacked(MerkleProofLib.getRoot(proof, _leaf), block.chainid));
 
-        address factory = _cachedThis;
-        assembly ("memory-safe") {
-            mstore(0x1d, 0x5af43d5f5f3e6022573d5ffd5b3d5ff3)
-            mstore(0x0d, factory)
-            mstore(0x00, 0x60265f8160095f39f35f5f365f5f37365f6c)
-            let initCodeHash := keccak256(0x0e, 0x2f)
-
-            // 0xff + factory + salt + hash(initCode)
-            mstore(0x00, 0xff00000000000000)
-            mstore(0x2d, salt)
-            mstore(0x4d, initCodeHash)
-            let computedAddress := keccak256(0x18, 0x55)
-
-            if shl(0x60, xor(computedAddress, address())) {
-                mstore(0x00, 0x00)
-                return(0x00, 0x20)
-            }
-            // Return ERC1271 magic value (isValidSignature selector)
-            mstore(0x00, 0x1626ba7e00000000000000000000000000000000000000000000000000000000)
-            return(0x00, 0x20)
-        }
+        _verifyRoot(MerkleProofLib.getRoot(proof, hash), pendingOwner());
```

Comment:
so if the ownership gets claimed, everything gets invalidated from that moment on. 👍

Reply:
yeah, I think that's a useful feature. of course, there's the weird edge case where if you `transferOwnership` back to the original owner, all of a sudden some signature again become valid. I think that I'm not gonna fix that, though